webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultCsvValidationService.java

```diff
@@ -0,0 +1,200 @@
+                        .build());
+                currentSchoolGroupYearSubject = subject;
+            }
+        }
+
+        return validationFailures;
+    }
+
+    private void validateSchoolGroupYear(final List<CsvValidationFailure> validationFailures,
+                                         final Set<String> completedSchoolGroupYears,
+                                         final String schoolGroupYear,
+                                         final long recordNumber) {
+        if (completedSchoolGroupYears.contains(schoolGroupYear)) {
+            validationFailures.add(CsvValidationFailure.builder()
+                    .row(recordNumber)
+                    .message("Record is out-of-order: sort CSV by school, group, year")
+                    .build());
+        }
+    }
+
+    @Override
+    public String toMessage(final List<CsvValidationFailure> failures) {
+        final StringBuilder messageBuilder = new StringBuilder();
+        int maxMessages = MaxFailureMessages;
+        for (final CsvValidationFailure failure : failures) {
+            messageBuilder.append(String.format(ValidationFormat, failure.getRow(), failure.getMessage()));
+            if (maxMessages-- < 0) {
+                messageBuilder.append(String.format(MoreFailuresFormat, failures.size() - MaxFailureMessages));
+                break;
+            }
+        }
+        return messageBuilder.toString();
+    }
+
+    private Integer getAsInt(final CSVRecord row, final String header) {
+        try {
+            return Integer.valueOf(row.get(header));
+        } catch (final NumberFormatException e) {
+            return null;
+        }
+    }
+
+    private void validateKeyValues(final List<CsvValidationFailure> validationFailures,
+                                                         final String school,
+                                                         final String group,
+                                                         final Integer schoolYear,
+                                                         final long recordNumber) {
+        if (isBlank(school)) {
+            validationFailures.add(CsvValidationFailure.builder()
+                    .row(recordNumber)
+                    .message("Record does not contain a school")
+                    .build());
+        }
+
+        if (isBlank(group)) {
+            validationFailures.add(CsvValidationFailure.builder()
+                    .row(recordNumber)
+                    .message("Record does not contain a group name")
+                    .build());
+        }
+
+        if (schoolYear == null) {
+            validationFailures.add(CsvValidationFailure.builder()
+                    .row(recordNumber)
+                    .message("Record does not contain a school year")
+                    .build());
+        }
+    }
+
+    private void validateSchool(final List<CsvValidationFailure> validationFailures,
```

Comment:
Hm... except that it's stateful validation, with validators modifying state and order being important.  That feels like it might reduce readability with hidden dependencies between validators.